### Import libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Import data

In [2]:
path = 'https://raw.githubusercontent.com/dan1dr/microdosing/main/data/dlongcodes.csv'
data = pd.read_csv(path, sep=',', index_col=False)
pd.options.display.max_columns = None

In [3]:
data.head(10)

ResponseId         StartDate           EndDate      RecordedDate  \
0  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
1  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
2  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
3  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
4  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
5  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
6  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   
7  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   
8  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   
9  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   

   MoodBoard_1  MoodBoard_2  MoodBoard_3  MoodBoard_4  MoodBoard_5  \
0            0            1            1            1            1   
1            0            1            1            1            1   
2            0            1            1            1            1   
3            0            1            1            1            1   
4            0            1            1            1            1   
5            0            1            1            1            1   
6            0            0            0            1            0   
7            0            0            0            1            0   
8            0            0            0            1            0   
9            0            0            0            1            0   

   MoodBoard_6  MoodBoard_7  MoodBoard_8  MoodBoard_9  MoodBoard_10  \
0            1            1            1            0             0   
1            1            1            1            0             0   
2            1            1            1            0             0   
3            1            1            1            0             0   
4            1            1            1            0             0   
5            1            1            1            0             0   
6            0            0            0            0             0   
7            0            0            0            0             0   
8            0            0            0            0             0   
9            0            0            0            0             0   

   MoodBoard_11  MoodBoard_12  MoodBoard_13  MoodBoard_14  MoodBoard_15  \
0             1             0             1             0             0   
1             1             0             1             0             0   
2             1             0             1             0             0   
3             1             0             1             0             0   
4             1             0             1             0             0   
5             1             0             1             0             0   
6             0             0             0             0             0   
7             0             0             0             0             0   
8             0             0             0             0             0   
9             0             0             0             0             0   

   MoodBoard_16  MoodBoard_17  MoodBoard_18  MoodBoard_19  MoodBoard_20  \
0             0             1             1             0             1   
1             0             1             1             0             1   
2             0             1             1             0             1   
3             0             1             1             0             1   
4             0             1             1             0             1   
5             0             1             1             0             1   
6             0             1             0             0             0   
7             0             1             0             0             0   
8             0             1             0             0             0   
9             0       

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1668 entries, 0 to 1667
Columns: 242 entries, ResponseId to ConCategory
dtypes: float64(138), int64(64), object(40)
memory usage: 3.1+ MB


In [5]:
data.shape

(1668, 242)

In [6]:
data.columns

Index(['ResponseId', 'StartDate', 'EndDate', 'RecordedDate', 'MoodBoard_1',
       'MoodBoard_2', 'MoodBoard_3', 'MoodBoard_4', 'MoodBoard_5',
       'MoodBoard_6',
       ...
       'ProRating', 'ResponseId.2', 'ConIndex', 'ConCode', 'Rating.C',
       'ConRating', 'ProConcept', 'ProCategory', 'ConConcept', 'ConCategory'],
      dtype='object', length=242)

## NAN's and duplicates

Since data comes from a survey conducted on the Internet, it may be sometimes be struggling to normalize and clean the format. Thankfully, this dataset was previously used and pre-cleaned by different authors https://osf.io/g5cwy/

However, the format and questionnaire performed is quite complex and we will need to make some adjustments.

Firstly, we need to assess the NANs, duplicates and other elements

In [7]:
def assess_NA(data):
    """
    Returns a pandas dataframe denoting the total number of NA
    values and the percentage of NA values in each column.
    The column names are noted on the index.
    """
    # pandas series denoting features and the sum of their null values
    null_sum = data.isnull().sum()# instantiate columns for missing data
    total = null_sum.sort_values(ascending=False)
    percent = ( ((null_sum / len(data.index))*100).round(2) ).sort_values(ascending=False)
    
    # concatenate along the columns to create the complete dataframe
    df_NA = pd.concat([total, percent], axis=1, sort=True, keys=['Number of NA', 'Percent NA'])
    
    # drop rows that don't have any missing data; omit if you want to keep all rows
    df_NA = df_NA[ (df_NA.T != 0).any() ]
    
    return df_NA
    
    
    

In [8]:
assess_NA(data)

Number of NA  Percent NA
BFI.2_63                            474       28.42
BFI.2_64                            474       28.42
BFI.2_65                            480       28.78
BFI.2_66                            474       28.42
BFI.2_67                            474       28.42
...                                 ...         ...
RedTobacco                           66        3.96
Reddit_5_TEXT...Topics             1308       78.42
Religion_12_TEXT...Topics          1602       96.04
SES                                  42        2.52
Substance_8_TEXT...Topics          1668      100.00

[160 rows x 2 columns]

Data is relatively "clean" and we will use all the rows given, most of NAs come from fields where the response ask for an alternative text in case of not being listed in the others before, so it is not relevant.

In [9]:
data['ResponseId'].nunique()

278

As we see, the data consists on a set of 278 participants identified by ``ResponseId``. The number of rows is 1668 because the table is given in a "long" format, where the pivot variable is the identity and repeats the same pattern every 6 rows (1668/6 = 278).

This will be important when dealing beyond the 'ProIndex' column, until that point we can dismiss this fact and work only with the unique values in order to acommodate the dataset.

Before splitting the dataset into two different parts, we will rename some columns in order to facilitate future tasks.

In [10]:
data = data.rename(columns={"MoodBoard_1": "Angry", "MoodBoard_2": "Distressed",
                    "MoodBoard_3": "Afraid", "MoodBoard_4": "Frustrated",
                    "MoodBoard_5": "Nervous", "MoodBoard_6": "Alarmed",
                    "MoodBoard_7": "Tense", "MoodBoard_8": "Annoyed",
                    "MoodBoard_9": "Astonished", "MoodBoard_10": "Excited",
                    "MoodBoard_11": "Alert", "MoodBoard_12": "Delighted",
                    "MoodBoard_13": "Amused", "MoodBoard_14": "Glad",
                    "MoodBoard_15": "Happy", "MoodBoard_16": "Pleased",
                    "MoodBoard_17": "Sad", "MoodBoard_18": "Miserable",
                    "MoodBoard_19": "Depressed", "MoodBoard_20": "Droopy",
                    "MoodBoard_21": "Gloomy", "MoodBoard_22": "Fatigued",
                    "MoodBoard_23": "Bored", "MoodBoard_24": "Tired",
                    "MoodBoard_25": "Serene", "MoodBoard_26": "Content",
                    "MoodBoard_27": "Calm", "MoodBoard_28": "At ease",
                    "MoodBoard_29": "Satisfied", "MoodBoard_30": "Relaxed",
                    "MoodBoard_31": "Sleepy", "MoodBoard_32": "Quiet"}, errors="raise")

Beyond the 'ProIndex' columns, we find the reason of the long format. Exploring the original data https://osf.io/dwtk5/ we find that there are represented the following values: the first three correspond to "pro_text_X_topics", and the last three to "pro_X_code". Being X from 1 to 3

The first 3 rows are almost identical to the following 3, just differ when sometimes one is NA or the topics is different from the category. We will select the category as it is a more general characteristic. Hence, we will only use the last 3 rows per each ``ResponseId``

We will divide the dataset into two parts in order to join them at the end
* ``wide_data``
* ``long_data``

### Wide info

In [11]:
wide_data = data.loc[:,:'DAS_Dep']
long_data = data.loc[:, 'ProIndex':]

In [12]:
wide_data = data.drop_duplicates(subset = 'ResponseId')
wide_data.shape

(278, 242)

In [13]:
wide_data.index

Int64Index([   0,    6,   12,   18,   24,   30,   36,   42,   48,   54,
            ...
            1608, 1614, 1620, 1626, 1632, 1638, 1644, 1650, 1656, 1662],
           dtype='int64', length=278)

In [14]:
#Rearrange the index from 0
wide_data.index = range(1,279)
wide_data.head()

ResponseId         StartDate           EndDate      RecordedDate  \
1  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
2  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   
3  R_0N8EpmmQ5b79cxr  18/10/2017 20:29  18/10/2017 23:51  18/10/2017 23:51   
4  R_0Nj0uZDZlPP0IRX  18/09/2017 11:40  18/09/2017 11:58  18/09/2017 11:58   
5  R_0OZWDFj2krlpms9  22/09/2017 18:03  23/09/2017 06:47  07/10/2017 06:47   

   Angry  Distressed  Afraid  Frustrated  Nervous  Alarmed  Tense  Annoyed  \
1      0           1       1           1        1        1      1        1   
2      0           0       0           1        0        0      0        0   
3      0           0       0           0        0        0      0        0   
4      0           0       0           0        0        0      0        0   
5      0           1       0           0        1        0      0        0   

   Astonished  Excited  Alert  Delighted  Amused  Glad  Happy  Pleased  Sad  \
1           0        0      1          0       1     0      0        0    1   
2           0        0      0          0       0     0      0        0    1   
3           0        0      0          0       0     0      1        0    0   
4           0        0      0          0       0     0      0        0    0   
5           0        1      1          0       1     0      1        1    0   

   Miserable  Depressed  Droopy  Gloomy  Fatigued  Bored  Tired  Serene  \
1          1          0       1       1         0      0      1       0   
2          0          0       0       0         1      1      1       0   
3          0          0       1       0         1      0      1       0   
4          0          0       1       0         1      0      0       0   
5          0          0       1       0         0      0      1       1   

   Content  Calm  At ease  Satisfied  Relaxed  Sleepy  Quiet  Current  Reddit  \
1        0     0        0          0        0       1      0        1       5   
2        1     1        0          0        1       0      1        1       1   
3        1     0        0          0        1       1      0        2       2   
4        0     0        0          0        0       0      0        1       1   
5        1     0        1          1        1       1      0        2       5   

   Country  Age  Gender  Orientation  Ethnicity Ethnicity_11_TEXT  Religion  \
1      187   21    -1.0         -1.0          1               -99         1   
2      187   29    -1.0         -1.0          6               -99         1   
3      187   20    -1.0         -1.0         13               -99         1   
4      128   31    -1.0         -1.0          1               -99         1   
5       -1   27    -1.0         -1.0          1               -99        14   

  Religion_12_TEXT  Religiosity  Spirituality  Education  SES  AgeMD  \
1              -99            2            20        6.0  1.0     20   
2              -99           14            19        7.0  2.0     28   
3              -99            0             5        3.0  2.0     19   
4              -99            0            28        6.0  1.0     32   
5              -99            2            68        6.0  2.0     26   

   Substance_1  Substance_2  Substance_8  Dose_1  Dose_4  Dose_2  Dose_3  \
1            1          -99          -99     -99       1     -99     -99   
2            1            1          -99     -99       1     -99     -99   
3            1          -99          -99       1       1     -99     -99   
4            1          -99          -99     -99       1     -99     -99   
5          -99            1          -99     -99     -99       1     -99   

   Dose_7  Dose_8  Spacing  DoseNumber  Money Money_1_TEXT Money_2_TEXT  \
1     -99     -99        3        25.0      1            5          -99   
2     -99     -99        1        26.0      1           20          -99   
3     -99     -99        5        25.0      1           10          -

### long_info

In [15]:
long_data.head(10)

ProIndex               ProCode       ResponseId.1  Rating.P  ProRating  \
0         1         Introspection  R_0HSeYpblz5LCLpD         1       85.0   
1         2  Perspective Shifting  R_0HSeYpblz5LCLpD         2       90.0   
2         3                   NaN  R_0HSeYpblz5LCLpD         3       55.0   
3         4         Introspection  R_0HSeYpblz5LCLpD         4       85.0   
4         5  Perspective Shifting  R_0HSeYpblz5LCLpD         5       90.0   
5         6             Self-Care  R_0HSeYpblz5LCLpD         6       55.0   
6         1            Creativity  R_0MpphsGkrwSMhuF         1       70.0   
7         2   Improved Motivation  R_0MpphsGkrwSMhuF         2       52.0   
8         3         Improved Mood  R_0MpphsGkrwSMhuF         3       45.0   
9         4            Creativity  R_0MpphsGkrwSMhuF         4       70.0   

        ResponseId.2  ConIndex             ConCode  Rating.C  ConRating  \
0  R_0HSeYpblz5LCLpD         1  Sleep Disturbances         1       25.0   
1  R_0HSeYpblz5LCLpD         2          Illegality         2        4.0   
2  R_0HSeYpblz5LCLpD         3              Stigma         3       20.0   
3  R_0HSeYpblz5LCLpD         4  Sleep Disturbances         4       25.0   
4  R_0HSeYpblz5LCLpD         5          Illegality         5        4.0   
5  R_0HSeYpblz5LCLpD         6              Stigma         6       20.0   
6  R_0MpphsGkrwSMhuF         1  Sleep Disturbances         1       35.0   
7  R_0MpphsGkrwSMhuF         2                 NaN         2        NaN   
8  R_0MpphsGkrwSMhuF         3                 NaN         3        NaN   
9  R_0MpphsGkrwSMhuF         4  Sleep Disturbances         4       35.0   

                ProConcept    ProCategory                  ConConcept  \
0                Self-Care  Self-Efficacy  Physiological Interference   
1  Meta-Creative Processes     Creativity                  Illegality   
2                      NaN            NaN                      Stigma   
3                Self-Care  Self-Efficacy  Physiological Interference   
4  Meta-Creative Processes     Creativity                  Illegality   
5                Self-Care  Self-Efficacy                      Stigma   
6               Creativity     Creativity  Physiological Interference   
7            Self-Efficacy  Self-Efficacy                         NaN   
8            Improved Mood  Improved Mood                         NaN   
9               Creativity     Creativity  Physiological Interference   

                ConCategory  
0  Physiological Discomfort  
1                Illegality  
2                Illegality  
3  Physiological Discomfort  
4                Illegality  
5                Illegality  
6  Physiological Discomfort  
7                       NaN  
8                       NaN  
9  Physiological Discomfort

In [16]:
long_data = long_data.drop(columns=['ResponseId.2','Rating.P'])
long_data.index = range(1,len(long_data)+1)

Now let's rearrange the columns to facilitate the comparison between pros and cons.

In [17]:
long_data.columns

Index(['ProIndex', 'ProCode', 'ResponseId.1', 'ProRating', 'ConIndex',
       'ConCode', 'Rating.C', 'ConRating', 'ProConcept', 'ProCategory',
       'ConConcept', 'ConCategory'],
      dtype='object')

In [18]:
def return_column_order(dataframe, order, rest=False, front=True):
    '''Takes a dataframe and reorder its columns. If "rest" is 
       True, it also takes into account the columns not listed, then 
       returns dataframe with order as first columns if "front" is 
       True, and order as last columns if "front" is False.
    '''
    cols = order[:]
    if rest:
        for x in dataframe.columns:
            if x not in cols:
                if front: #insert in the front in the list
                    cols.append(x)
                else: #insert to be in the last
                    cols.insert(0, x)
    return dataframe[cols]

In [19]:
long_data = return_column_order(long_data, ['ResponseId.1', 'ProCode', 'ProCategory',
                        'ProRating', 'ConCode', 'ConCategory',
                        'ConRating'])

Now we want to pivot from the "RespondeId.1" and transform it to a wide data. We will only use the last 3 rows per respondent as discussed before. So previous to pivoting let's select 3 rows per user and later create ProCodeX, ProCategory and ProRatingX for i from 1 to 3. Same proccess with Cons.

In [20]:
long_data

ResponseId.1               ProCode         ProCategory  ProRating  \
1     R_0HSeYpblz5LCLpD         Introspection       Self-Efficacy       85.0   
2     R_0HSeYpblz5LCLpD  Perspective Shifting          Creativity       90.0   
3     R_0HSeYpblz5LCLpD                   NaN                 NaN       55.0   
4     R_0HSeYpblz5LCLpD         Introspection       Self-Efficacy       85.0   
5     R_0HSeYpblz5LCLpD  Perspective Shifting          Creativity       90.0   
...                 ...                   ...                 ...        ...   
1664  R_ZfYg5JxACJX7szf    Clarity of Thought  Cognitive Benefits       79.0   
1665  R_ZfYg5JxACJX7szf         Improved Mood       Improved Mood      100.0   
1666  R_ZfYg5JxACJX7szf           Sociability     Social Benefits       49.0   
1667  R_ZfYg5JxACJX7szf    Clarity of Thought  Cognitive Benefits       79.0   
1668  R_ZfYg5JxACJX7szf         Improved Mood       Improved Mood      100.0   

                          ConCode                ConCategory  ConRating  
1              Sleep Disturbances   Physiological Discomfort       25.0  
2                      Illegality                 Illegality        4.0  
3                          Stigma                 Illegality       20.0  
4              Sleep Disturbances   Physiological Discomfort       25.0  
5                      Illegality                 Illegality        4.0  
...                           ...                        ...        ...  
1664                          NaN                        NaN       59.0  
1665                          NaN                        NaN        9.0  
1666                         None  Other Perceived Drawbacks        7.0  
1667  Unknown Risk-Effect Profile  Other Perceived Drawbacks       59.0  
1668                          NaN                        NaN        9.0  

[1668 rows x 7 columns]

In [21]:
type(len(long_data)/long_data['ResponseId.1'].nunique())

float

In [22]:
list = []
    #step in range must be converted into a integer
for i in range(3, len(long_data), int(len(long_data)/long_data['ResponseId.1'].nunique())):
    list.append(i)
    list.append(i+1)
    list.append(i+2)

In [23]:
def select_last3(dataframe):
    '''
    Selects from the dataframe the last 3 rows per ResponseId
    '''
    list = []
    #step in range must be converted into a integer
    for i in range(3, len(long_data), int(len(long_data)/long_data['ResponseId.1'].nunique())):
        list.append(i)
        list.append(i+1)
        list.append(i+2)
    return dataframe.iloc[list]

In [24]:
long_data = select_last3(long_data)

In [25]:
long_data

ResponseId.1               ProCode         ProCategory  ProRating  \
4     R_0HSeYpblz5LCLpD         Introspection       Self-Efficacy       85.0   
5     R_0HSeYpblz5LCLpD  Perspective Shifting          Creativity       90.0   
6     R_0HSeYpblz5LCLpD             Self-Care       Self-Efficacy       55.0   
10    R_0MpphsGkrwSMhuF            Creativity          Creativity       70.0   
11    R_0MpphsGkrwSMhuF   Improved Motivation       Self-Efficacy       52.0   
...                 ...                   ...                 ...        ...   
1661  R_yUcpPqMIp5DPBSN       Improved Energy     Improved Energy       80.0   
1662  R_yUcpPqMIp5DPBSN        Improved Focus      Improved Focus      100.0   
1666  R_ZfYg5JxACJX7szf           Sociability     Social Benefits       49.0   
1667  R_ZfYg5JxACJX7szf    Clarity of Thought  Cognitive Benefits       79.0   
1668  R_ZfYg5JxACJX7szf         Improved Mood       Improved Mood      100.0   

                          ConCode                ConCategory  ConRating  
4              Sleep Disturbances   Physiological Discomfort       25.0  
5                      Illegality                 Illegality        4.0  
6                          Stigma                 Illegality       20.0  
10             Sleep Disturbances   Physiological Discomfort       35.0  
11                            NaN                        NaN        NaN  
...                           ...                        ...        ...  
1661                  Preparation  Other Perceived Drawbacks        5.0  
1662                         Misc  Other Perceived Drawbacks        5.0  
1666                         None  Other Perceived Drawbacks        7.0  
1667  Unknown Risk-Effect Profile  Other Perceived Drawbacks       59.0  
1668                          NaN                        NaN        9.0  

[834 rows x 7 columns]

In [26]:
long_data['ResponseId.1'].nunique()

278

Another strategy could have been using the 'ProIndex' column previously deleted and selecting the values from 4 to 6 per ``ResponseId.1``.

Now we have all we need to pivot the data, so we create the second DataFrame we want to input the long info. 

In [30]:
df = pd.DataFrame(index=range(1,279),
                  columns= ['ResponseId', 'ProCode1', 'ProCategory1', 'ProRating1',
                 'ConCode1', 'ConCategory1', 'ConRating1',
                'ProCode2', 'ProCategory2', 'ProRating2',
                 'ConCode2', 'ConCategory2', 'ConRating2',
                'ProCode3', 'ProCategory3', 'ProRating3',
                 'ConCode3', 'ConCategory3', 'ConRating3'])
df['ResponseId'] = wide_data['ResponseId']
# This dataframe should be equal to the number of rows to the initial table with the uniques values.  
len(df) == len(wide_data)

True

Now we need to create 3 sequence lists (i.e. The 4th, 5th, and 6th row per respondent and later added it to the corrrespondent column)

In [31]:
list1 = []
for i in range(0, len(long_data), int(len(long_data)/long_data['ResponseId.1'].nunique())):
        list1.append(i)
list2 = []
for j in range(1, len(long_data), int(len(long_data)/long_data['ResponseId.1'].nunique())):
        list2.append(j)
list3 = []
for k in range(2, len(long_data), int(len(long_data)/long_data['ResponseId.1'].nunique())):
        list3.append(k)

We use each list to filter each sequences and added it to the corresponding columns created in df

Time to assemble the three lists to the columns created.

In [32]:
long_data.columns

Index(['ResponseId.1', 'ProCode', 'ProCategory', 'ProRating', 'ConCode',
       'ConCategory', 'ConRating'],
      dtype='object')

In [33]:
df.columns

Index(['ResponseId', 'ProCode1', 'ProCategory1', 'ProRating1', 'ConCode1',
       'ConCategory1', 'ConRating1', 'ProCode2', 'ProCategory2', 'ProRating2',
       'ConCode2', 'ConCategory2', 'ConRating2', 'ProCode3', 'ProCategory3',
       'ProRating3', 'ConCode3', 'ConCategory3', 'ConRating3'],
      dtype='object')

In [34]:
long_data.iloc[list1]['ProCode']

4            Introspection
10              Creativity
16         Improved Energy
22      Clarity of Thought
28              Creativity
               ...        
1642                  Misc
1648           Sociability
1654            Creativity
1660    Clarity of Thought
1666           Sociability
Name: ProCode, Length: 278, dtype: object

In [35]:
def input_data(old, new):
    '''
    Inputs the columns from the old dataframe to the new one.
    First filters the old dataframe with the index, equals its index
    to the new and then it is added each column
    '''
    #long_data.reset_index(drop=True, inplace=True)
    #long_data.drop(columns=['level_0','index'])
    for i in range(1,4):
        if i == 1:
            list = list1
            old1 = old.iloc[list]
            old1.index = new.index
            new['ProCode1'] = old1['ProCode']
            new['ProCategory1'] = old1['ProCategory']
            new['ProRating1'] = old1['ProRating']
            new['ConCode1'] = old1['ConCode']
            new['ConCategory1'] = old1['ConCategory']
            new['ConRating1'] = old1['ConRating']
            
        elif i == 2:
            list = list2
            old2 = old.iloc[list]
            old2.index = new.index
            new['ProCode2'] = old2['ProCode']
            new['ProCategory2'] = old2['ProCategory']
            new['ProRating2'] = old2['ProRating']  
            new['ConCode2'] = old2['ConCode']
            new['ConCategory2'] = old2['ConCategory']
            new['ConRating2'] = old2['ConRating']
            
        elif i == 3:
            list = list3
            old3 = old.iloc[list]
            old3.index = new.index
            new['ProCode3'] = old3['ProCode']
            new['ProCategory3'] = old3['ProCategory']
            new['ProRating3'] = old3['ProRating']  
            new['ConCode3'] = old3['ConCode']
            new['ConCategory3'] = old3['ConCategory']
            new['ConRating3'] = old3['ConRating']
            
        else:
            break
    return new

In [36]:
wide_data2 = input_data(long_data,df)
wide_data2.shape == df.shape

True

In [37]:
wide_data2

ResponseId            ProCode1              ProCategory1  \
1    R_0HSeYpblz5LCLpD       Introspection             Self-Efficacy   
2    R_0MpphsGkrwSMhuF          Creativity                Creativity   
3    R_0N8EpmmQ5b79cxr     Improved Energy           Improved Energy   
4    R_0Nj0uZDZlPP0IRX  Clarity of Thought        Cognitive Benefits   
5    R_0OZWDFj2krlpms9          Creativity                Creativity   
..                 ...                 ...                       ...   
274  R_xDYH5RjCj17uTyp                Misc  Other Perceived Benefits   
275  R_Y03YO4iKPXHfyOl         Sociability           Social Benefits   
276  R_Y98iLISR4MDPZVT          Creativity                Creativity   
277  R_yUcpPqMIp5DPBSN  Clarity of Thought        Cognitive Benefits   
278  R_ZfYg5JxACJX7szf         Sociability           Social Benefits   

     ProRating1                     ConCode1               ConCategory1  \
1          85.0           Sleep Disturbances   Physiological Discomfort   
2          70.0           Sleep Disturbances   Physiological Discomfort   
3          90.0              Body Discomfort   Physiological Discomfort   
4          97.0                          NaN                        NaN   
5          99.0                      Anxiety          Increased Anxiety   
..          ...                          ...                        ...   
274        50.0  Unknown Risk-Effect Profile  Other Perceived Drawbacks   
275       100.0               Visual Effects   Physiological Discomfort   
276       100.0                   Illegality                 Illegality   
277       100.0              Body Discomfort   Physiological Discomfort   
278        49.0                         None  Other Perceived Drawbacks   

     ConRating1               ProCode2        ProCategory2  ProRating2  \
1          25.0   Perspective Shifting          Creativity        90.0   
2          35.0    Improved Motivation       Self-Efficacy        52.0   
3          20.0             Creativity          Creativity        65.0   
4           NaN  Improved Productivity       Self-Efficacy       100.0   
5          34.0   Perspective Shifting          Creativity        99.0   
..          ...                    ...                 ...         ...   
274       100.0                    NaN                 NaN         NaN   
275        15.0          Improved Mood       Improved Mood       100.0   
276        34.0  Cognitive Enhancement  Cognitive Benefits       100.0   
277         5.0        Improved Energy     Improved Energy        80.0   
278         7.0     Clarity of Thought  Cognitive Benefits        79.0   

                        ConCode2               ConCategory2  ConRating2  \
1                     Illegality                 Illegality         4.0   
2                            NaN                        NaN         NaN   
3             Sleep Disturbances   Physiological Discomfort        65.0   
4                            NaN                        NaN         NaN   
5                  Dose Accuracy                 Illegality        32.0   
..                           ...                        ...         ...   
274                          NaN                        NaN         NaN   
275                  Preparation  Other Perceived Drawbacks        50.0   
276              Body Discomfort   Physiological Discomfort        15.0   
277                  Preparation  Other Perceived Drawbacks         5.0   
278  Unknown Risk-Effect Profile  Other Perceived Drawbacks        59.0   

                  ProCode3    ProCategory3  ProRating3         ConCode3  \
1                Self-Care   Self-Efficacy        55.0           Stigma   
2            Improved Mood   Improved Mood        45.0              NaN   
3            Improved Mood   Improved Mood        72.0  Racing Thoughts   
4           Improved Focus  Improved Focus       100.0              NaN   
5                 Openness      Creativity        98.0              NaN   
..    

## Merge data for final use

In [38]:
wide = wide_data.merge(wide_data2, on='ResponseId')
#pd.concat([wide_data,wide_data2], axis=1)

In [40]:
wide

ResponseId         StartDate           EndDate      RecordedDate  \
0    R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
1    R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   
2    R_0N8EpmmQ5b79cxr  18/10/2017 20:29  18/10/2017 23:51  18/10/2017 23:51   
3    R_0Nj0uZDZlPP0IRX  18/09/2017 11:40  18/09/2017 11:58  18/09/2017 11:58   
4    R_0OZWDFj2krlpms9  22/09/2017 18:03  23/09/2017 06:47  07/10/2017 06:47   
..                 ...               ...               ...               ...   
273  R_xDYH5RjCj17uTyp  12/09/2017 23:26  12/09/2017 23:48  26/09/2017 23:48   
274  R_Y03YO4iKPXHfyOl  07/10/2017 11:13  07/10/2017 11:22  21/10/2017 11:22   
275  R_Y98iLISR4MDPZVT  30/09/2017 14:20  30/09/2017 14:42  14/10/2017 14:42   
276  R_yUcpPqMIp5DPBSN  19/09/2017 11:13  19/09/2017 11:25  03/10/2017 11:25   
277  R_ZfYg5JxACJX7szf  18/09/2017 14:22  18/09/2017 14:40  18/09/2017 14:40   

     Angry  Distressed  Afraid  Frustrated  Nervous  Alarmed  Tense  Annoyed  \
0        0           1       1           1        1        1      1        1   
1        0           0       0           1        0        0      0        0   
2        0           0       0           0        0        0      0        0   
3        0           0       0           0        0        0      0        0   
4        0           1       0           0        1        0      0        0   
..     ...         ...     ...         ...      ...      ...    ...      ...   
273      0           0       0           1        0        0      1        1   
274      0           0       0           0        0        0      0        1   
275      0           0       0           0        0        0      1        1   
276      0           0       0           0        0        0      0        1   
277      0           0       0           0        0        0      0        0   

     Astonished  Excited  Alert  Delighted  Amused  Glad  Happy  Pleased  Sad  \
0             0        0      1          0       1     0      0        0    1   
1             0        0      0          0       0     0      0        0    1   
2             0        0      0          0       0     0      1        0    0   
3             0        0      0          0       0     0      0        0    0   
4             0        1      1          0       1     0      1        1    0   
..          ...      ...    ...        ...     ...   ...    ...      ...  ...   
273           0        0      0          0       0     0      0        0    0   
274           0        0      0          0       1     1      1        1    0   
275           0        0      0          0       0     0      0        0    0   
276           0        1      0          1       1     1      1        1    0   
277           0        0      0          0       0     0      0        0    0   

     Miserable  Depressed  Droopy  Gloomy  Fatigued  Bored  Tired  Serene  \
0            1          0       1       1         0      0      1       0   
1            0          0       0       0         1      1      1       0   
2            0          0       1       0         1      0      1       0   
3            0          0       1       0         1      0      0       0   
4            0          0       1       0         0      0      1       1   
..         ...        ...     ...     ...       ...    ...    ...     ...   
273          0          0       1       1         1      1      1       0   
274          0          0       0       0         1      1      0       0   
275          0          0       0       0         1      1      1       0   
276          0          0       0       0         0      1      1       1   
277          0          0       0       0         1      0      0       0   

     Content  Calm  At ease  Satisfied  Relaxed  Sleepy  Quiet  Current  \
0          0     0        0          0        0       1      0        1   
1          1     1        0          0    

In [41]:
wide.to_csv('./wide.csv', sep=",")